department-of-veterans-affairs/caseflow
# Fix Email Fallback and Duplicate/Orphan HearingEmailRecipients

### Description
This code fixes two classes of problem with our email recipient logic. #2 was only testable and solvable after fixing #1, which is why they're both included in this PR.

### The Problem Summary:
1. Email Address Fallbacks: In master currently, if you convert (or schedule?) a hearing to virtual without entering an email address, it will fall back to an email address on the veteran record. This means practically:
    - There are situations where there would be a HearingEmailRecipient for an email that was not typed in by the coordinator at time of scheduling.
    - Our test data always has the email on the veteran, prod does not, so our tests were not reflecting the reality of prod.  
2. Duplicate Hearing Email Recipients: In master currently, converting any `Legacy` hearing (Travel, or Video, or Central) to virtual where the Legacy hearing doesn't already have appellant and representative emails would cause that hearing to get 2 AppellantHearingEmailRecipients and 2 RepresentativeHearingEmailRecipients.
    - This is incorrect, we should only have one of each of these types per hearing.
    - This also puts the hearing in a 'stuck' state where it's not possible to edit the email/timezone for the appellant or representative recipients.
    - I expect to followup with a data cleanup PR.

### Part 1 - Email Address Fallbacks
We expect a hearing to have HearingEmailRecipient if and only if the user has entered an email. Right now, we go and get an email from an external source even if they don't enter an email (see testing for reproduction)

The Code Changed:
- Everything in this PR that's not in `create_or_update_recipients`

Acceptance Criteria
- [ ] It's possible to schedule a hearing and view it without any HearingEmailRecipients appearing.

Testing Plan
- [ ] First, look at the existing behavior in master. Do this with an (AMA) hearing to avoid hitting the other bug.
    - `git checkout master`
    - Schedule a veteran, but don't enter an email address or timezone for either appellant or representative.
    - Once you've clicked "Schedule" refresh the page you land on (appeal details) so the hearing appears.
    - Go find the "View Hearing Details" link on that page, but don't click it. Get the hearing uuid out of that link.
    - Switch to the console now (without clicking that details link!)
    - `h = Hearing.find_by_uuid("[uuid from the link]")`
    - `h.email_recipients.count # This should be 0`
    - Okay, now in the UI click the "View Hearing Details" link and run in the console
    - `h = Hearing.find_by_uuid("[uuid from the link]")`
    - `h.email_recipients.count # This now shows 2/3
    - Note that viewing the hearing details page create HearingEmailRecipients. This is incorrect behavior.
- [ ] Now checkout this branch and repeat those steps. You should not see any HearingEmailRecipients created in the last few steps. Accessing the View Hearing Details link does NOT create HearingEmailRecipients any longer.

### Part 2 - Duplicate Hearing Email Recipients
For Legacy Hearings that don't have appellant and representative emails/timezones, converting them to virtual and entering an email/timezone (required) will put the hearing in a stuck state right now. You can see the stuck state on the hearing details page, the email and timezone fields are grayed out.

The Code Changed:
- Look up recipients differently in `create_or_update_recipients`

Acceptance Criteria
- [ ] You can convert a LegacyHearing to Virtual and
    - [ ] It only has one AppellantHearingEmailRecipient and one RepresentativeHearingEmailRecipient.
    - [ ] The UI shows the emails and timezones as editable.

Testing Plan
- [ ] Do the following on `master` branch to see the problem, then check out this branch and do it again to verify the problem goes away.
    - `git checkout master`
    - Go to the hearing schedule, find a travel day (or legacy video, but travel is easier to find in the UI)
    - For that travel day "Edit Hearing Details" for one of the travel hearings. Make sure the bottom of the screen shows no emails entered.
    - Click the "Type" dropdown and select "Virtual".
    - Fill in the appellant email like: "trevor@example.com"
    - Fill in the appellant timezone like "Eastern"
    - Fill in the representative email like "trevor-rep@example.com"
    - Fill in the representative timezone like "pacific"
    - Choose a time if required
    - Click "Schedule"
    - Note that if you're on `master` branch the next screen will not show the representative email and the email/timezone fields will be grayed out. This should be fixed on this branch
    - Also go into the console, and do this (on both master and branch)
``` 
lh = LegacyHearing.find_by(vacols_id: 410006)
lh.email_recipients # There should NOT be two of each type (rep and appellant)
```
